File: backup/lib/bannered_introduction.rb
class BanneredIntroduction
  attr_reader :words, :banner_nodes, :colors

  def initialize(words)
    @words = words.split(" ")
    @colors = ["red", "light blue", "orange"]
    create_headered_span_words
  end

  def create_headered_span_words(header_size = 2)
    nodes = words.map do |word|
      chars         = word.split("")
      spans         = chars.map.with_index { |char, i| span_for_character(char, chars, i) }.join
      headered_span = "<h#{header_size}>#{spans}</h#{header_size}>"
    end

    @banner_nodes = nodes
  end

  def span_for_character(char, chars, index)
    if index == chars.length - 1
      "<span style='color: #{colors.sample}'>#{char}</br></span>\n"
    else
      "<span style='color: #{colors.sample}'>#{char}</span>\n"
    end
  end
end

File: elixir_onepg/lib/onepg_ex/bannered_introduction.ex
defmodule OnepgEx.BanneredIntroduction do
  defstruct nodes: [], body: "", words: []

  @colors ["red", "light blue", "orange"]

  def create_headered_span_words(words, header_size) do
    nodes =
      words
      |> String.split(" ")
      |> Enum.map(fn word ->
        chars = String.split(word, "")
        spans = spans_for_words(chars)

        "<h#{header_size}>#{spans}</h#{Integer.to_string(header_size)}>"
      end)

    html = Enum.join(nodes, "\n")
    %Onepg.HtmlDocument{body: html, nodes: nodes}
  end

  def spans_for_words(chars) do
    chars
    |> Enum.with_index()
    |> Enum.map(fn {char, i} ->
      span_for_character(char, chars, i)
    end)
  end

  def span_for_character(char, chars, index) do
    [color, _, _] = Enum.shuffle(@colors)
    is_last = index == Enum.count(chars) - 1

    case is_last do
      true -> "<span style='color: #{color}'>#{char}</br></span>\n"
      false -> "<span style='color: #{color}'>#{char}</span>\n"
    end
  end
end
